tests/test_bootstrap.py

```diff
@@ -0,0 +1,551 @@
+
+        # test that prepare_build_dir runs (notice that we mock
+        # any file/dir creation so we can speed up the tests)
+        bs.prepare_build_dir()
+        # make sure that the open command has been called only once
+        mock_open.assert_called_once_with("project.properties", "w")
+        mock_open.reset_mock()
+
+    @mock.patch("pythonforandroid.bootstrap.os.path.isfile")
+    @mock.patch("pythonforandroid.bootstrap.os.path.exists")
+    @mock.patch("pythonforandroid.bootstrap.os.unlink")
+    @mock.patch("pythonforandroid.bootstrap.open", create=True)
+    @mock.patch("pythonforandroid.util.chdir")
+    @mock.patch("pythonforandroid.bootstrap.shutil.copy")
+    @mock.patch("pythonforandroid.bootstrap.os.makedirs")
+    @mock.patch("pythonforandroid.bootstrap.sh.ln")
+    @mock.patch("pythonforandroid.bootstrap.listdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.mkdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.rm")
+    @mock.patch("pythonforandroid.bootstrap.isdir")
+    def test_bootstrap_prepare_build_dir_with_java_src(
+        self,
+        mock_isdir,
+        mock_sh_rm,
+        mock_sh_mkdir,
+        mock_listdir,
+        mock_sh_ln,
+        mock_os_makedirs,
+        mock_shutil_copy,
+        mock_chdir,
+        mock_open,
+        mock_os_unlink,
+        mock_os_path_exists,
+        mock_os_path_isfile,
+    ):
+        """A test which will initialize a bootstrap and will check perform
+        another test for method
+        :meth:`~pythonforandroid.bootstrap.Bootstrap.prepare_build_dir`. In
+        here we will simulate that we have `with_java_src` set to some value.
+        """
+        self.ctx.symlink_java_src = ["some_java_src"]
+        mock_isdir.return_value = False
+        mock_sh_mkdir.return_value = "sh.mkdir successfully mocked"
+        mock_sh_rm.return_value = "sh.rm successfully mocked"
+        mock_listdir.return_value = [
+            "jnius",
+            "kivy",
+            "Kivy-1.11.0.dev0-py3.7.egg-info",
+            "pyjnius-1.2.1.dev0-py3.7.egg",
+        ]
+        mock_sh_ln.return_value.return_value = mock.Mock()
+        mock_os_makedirs.return_value = True
+        mock_shutil_copy.return_value = True
+        mock_chdir.return_value = True
+        mock_open.side_effect = [
+            mock.mock_open(read_data="target=android-21").return_value
+        ]
+        mock_os_unlink.return_value = True
+        mock_os_path_exists.return_value = True
+        mock_os_path_isfile.return_value = True
+
+        # prepare bootstrap
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.ctx.bootstrap = bs
+
+        # test that prepare_build_dir runs (notice that we mock
+        # any file/dir creation so we can speed up the tests)
+        bs.prepare_build_dir()
+        # make sure that the open command has been called only once
+        mock_open.assert_called_with("project.properties", "w")
+        mock_open.reset_mock()
```

Comment:
Minor: Maybe we can also quick check that the symlink was made 4 times since we listdir returned 4 dirs?
E.g. something like:
```python
self.assertEqual(len(mock_sh_ln.call_args_list), len(mock_listdir.return_value)
```
Or if we want to go a bit further, we could check the symlink source, e.g.
```python
for i, directory in enumerate(mock_listdir.return_value):
    self.assertTrue(mock_sh_ln.call_args_list[i][0][1].endswith(directory))
```

Reply:
Thanks @AndreMiras :smile:
I introduced your suggestion, if you think that we should add more tests let me know (actually I'm pretty sure that we can test a lot more, but for starters it's fine...we got almost all code executed and we improve our coverage...)

Reply:
Yes I definitely think what you had was more than good for a start :+1: